src/engine/GroupBy.cpp

```diff
@@ -388,15 +373,125 @@ ProtoResult GroupBy::computeResult([[maybe_unused]] bool requestLaziness) {
     return {std::move(idTable), resultSortedOn(), std::move(localVocab)};
   }
 
-  size_t inWidth = subresult->idTable().numColumns();
-  size_t outWidth = getResultWidth();
+  // TODO<RobinTF> Add runtime parameter to toggle of index scan specific
+  // optimizations for performance comparisons
+  if (subresult->isFullyMaterialized()) {
+    size_t inWidth = subresult->idTable().numColumns();
+    size_t outWidth = getResultWidth();
+
+    IdTable idTable = CALL_FIXED_SIZE(
+        (std::array{inWidth, outWidth}), &GroupBy::doGroupBy, this,
+        subresult->idTable(), groupByCols, aggregates, &localVocab);
+
+    LOG(DEBUG) << "GroupBy result computation done." << std::endl;
+    return {std::move(idTable), resultSortedOn(), std::move(localVocab)};
+  }
 
-  IdTable idTable = CALL_FIXED_SIZE(
-      (std::array{inWidth, outWidth}), &GroupBy::doGroupBy, this,
-      subresult->idTable(), groupByCols, aggregates, &localVocab);
+  auto aggregateInfo = computeHashMapOptimizationMetadata(aggregates);
 
-  LOG(DEBUG) << "GroupBy result computation done." << std::endl;
-  return {std::move(idTable), resultSortedOn(), std::move(localVocab)};
+  if (!aggregateInfo.has_value()) {
+    // TODO<RobinTF> Should this throw an exception?
+    return {IdTable{getResultWidth(), getExecutionContext()->getAllocator()},
+            resultSortedOn(), std::move(localVocab)};
+  }
```

Comment:
No?
You should use the fully materialized fallback for this.
So first (before doing anything) check if you can do lazy with the given aggregates, else set the `requestLazy` Parameter to false.